src/main/java/build/buildfarm/worker/resources/LocalResourceSetUtils.java

```diff
@@ -37,50 +43,95 @@
 public class LocalResourceSetUtils {
   private static final LocalResourceSetMetrics metrics = new LocalResourceSetMetrics();
 
+  private record SemaphoreLease(String name, int amount, Claim.Stage stage) {}
+
+  private record PoolLease(String name, List<Object> claims, Claim.Stage stage) {}
+
   public static final class LocalResourceSetClaim implements Claim {
-    private final List<Entry<String, Integer>> semaphores;
-    @Getter private final List<Entry<String, List<Integer>>> pools;
+    private final List<SemaphoreLease> semaphores;
+    private final List<PoolLease> pools;
     private final LocalResourceSet resourceSet;
 
     LocalResourceSetClaim(
-        List<Entry<String, Integer>> semaphores,
-        List<Entry<String, List<Integer>>> pools,
-        LocalResourceSet resourceSet) {
+        List<SemaphoreLease> semaphores, List<PoolLease> pools, LocalResourceSet resourceSet) {
       this.semaphores = semaphores;
       this.pools = pools;
       this.resourceSet = resourceSet;
     }
 
+    @Override
+    public synchronized void release(Stage stage) {
+      // only remove resources that match the specified stage
+      // O(n) search for staged resources
+      Iterator<SemaphoreLease> semaphoreIter = semaphores.iterator();
+      while (semaphoreIter.hasNext()) {
+        SemaphoreLease semaphore = semaphoreIter.next();
```

Comment:
suggestion: is there a more modern way to do this? :) Streams?

Reply:
Not with the removal that I can see. I'd change the reference if I did a filter, not sure if there's a stream-y predicated remover.